### PR TITLE
Add a TypeScript shareable configuration.

### PR DESCRIPTION
**Pull request by TrevorKarjanis**
- Add a TypeScript shareable configuration.

Notes
- All of the extension rule configurations are based on their core values in the JavaScript configuration.
- The [parserOptions.project](https://www.npmjs.com/package/@typescript-eslint/parser#parseroptionsproject) configuration is for rules that require type information.
- The [process for setting up an Angular workspace](https://github.com/angular-eslint/angular-eslint#migrating-an-angular-cli-project-from-codelyzer-and-tslint) is different and will be added later.

Linted and tested
- Verified that the configuration works with a standard ESLint workspace
- Verified that the configuration works with an Angular workspace

### CODE REVIEW

File: typescript.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
+module.exports = {
+    extends: './index',
+    parser: '@typescript-eslint/parser',
+    rules: {
```
Comment (jattasNI) [inline typescript.js:4]: I don't quite follow where this initial set of rules came from. At first I thought it was our JS rules in `index.js` but I don't think that's the case because I see lots of differences. Can someone help me understand?

**Comment by jattasNI** [comment]
Our test suite currently invokes ESLint on the JS files in the repo. Would there be value in adding a placeholder TS file and linting it too? Might be overkill, just a suggestion.

File: typescript.js (hunk 1 above)
Comment (TrevorKarjanis) [reply typescript.js:4]: They are from [here](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin#extension-rules). Their configurations should be equivalent to our ```index.js``` or the airbnb base. Can you point out an example of an incorrect one? Maybe I messed up or we're comparing differently. For example, brace-style is from [here](https://github.com/airbnb/javascript/blob/63098cbb6c05376dbefc9a91351f5727540c1ce1/packages/eslint-config-airbnb-base/rules/style.js#L21) while comma-dangle is from [here](https://github.com/ni/javascript-styleguide/blob/4e4d8426f1a6269f29c99682aade36538f1cedd1/index.js#L21).

Comment (jattasNI) [reply typescript.js:4]: Ah, I understand better now. It was just confusion on my part. I didn't notice that this extends `index.js`. So I thought there would be a 1-1 mapping from rules in our `index.js` to rules in this file and I got confused when scanning and saw that some rules from `index.js` didn't appear here at all (which I now understand is because the config in `index` can apply to TypeScript code unchanged) and some rules appeared here multiple times (which I now understand is to turn off the original and to replace it with one from @typescript-eslint).

After wrapping my head around it, I didn't find any examples configured incorrectly (but I did more of a spot check than a comprehensive audit).

**Review by jattasNI — APPROVED** [review]

**Comment by TrevorKarjanis** [comment]
> Our test suite currently invokes ESLint on the JS files in the repo. Would there be value in adding a placeholder TS file and linting it too? Might be overkill, just a suggestion.

Yes, I think even a smoke test is valuable. However, that would require a TypeScript configuration and test file which I would like to leave for another PR. This ruleset currently requires type checking whereas @typescript-eslint/recommended separates rules that require types. That is a [different discussion](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/TYPED_LINTING.md#performance) altogether.

File: typescript.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (TrevorKarjanis) [reply typescript.js:4]: That is a great but complicated question with a twisted answer.

TL;DR No, as with @typescript-eslint/recommended we want to [support unknown file types](https://github.com/typescript-eslint/typescript-eslint/pull/741#issuecomment-552694613). Overrides are, ["applied only if the glob patterns of both of the parent config and the child config [match],"](https://eslint.org/docs/user-guide/configuring#how-it-works) and would restrict usage to the defined extensions.

The @typescript-eslint/eslint-recommended ruleset is, ["meant to be used after extending eslint:recommended."](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin/src/configs#eslint-recommended) They are not extension rules, and I think the override is used to ensure they are only applied where the TypeScript compiler would be used, e.g. .ts files but not markdown, etc.

There is also the question of if the TypeScript configuration should ever be run on JavaScript files. [There are TypeScript rules that are not compatible with JavaScript, and @typescript-eslint recommends using overrides.](https://github.com/typescript-eslint/typescript-eslint/issues/109#issuecomment-530909615) There are also performance issues when using rules that require type checking

[the remark this answers is not in the document]
Comment (TrevorKarjanis) [reply typescript.js:4]: That is a great but complicated question with a twisted answer.

TL;DR No, as with @typescript-eslint/recommended we want to [support unknown file types](https://github.com/typescript-eslint/typescript-eslint/pull/741#issuecomment-552694613). Overrides are, ["applied only if the glob patterns of both of the parent config and the child config [match],"](https://eslint.org/docs/user-guide/configuring#how-it-works) and would restrict usage to the defined extensions.

The @typescript-eslint/eslint-recommended ruleset is, ["meant to be used after extending eslint:recommended."](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin/src/configs#eslint-recommended) They are not extension rules, and I think the override is used to ensure they are only applied where the TypeScript compiler would be used, e.g. .ts files but not markdown, etc.

There is then the question of if the TypeScript configuration should ever be run on JavaScript files. [There are a few TypeScript rules that are not compatible with JavaScript, and @typescript-eslint recommends using overrides.](https://github.com/typescript-eslint/typescript-eslint/issues/109#issuecomment-530909615) It may be reasonable to advise that the TypeScript configuration can be run on JavaScript if there aren't many problematic rules in our final configuration. However, if the workspace has many JavaScript files that are not included in the compiler configuration then they may require an override to avoid [performance problems](https://github.com/angular-eslint/angular-eslint#eslint-configs-and-performance).

At the least, I think it's reasonable to include an explanation and example of linting a mixed workspace with overrides. We can make a recommendation on when to use each approach once we have a configuration.

[the remark this answers is not in the document]
Comment (LukasKall) [reply typescript.js:4]: I see, that makes sense. I was just curious because I saw the usage of overrides at eslint-recommended. I agree that we should focus on the TypeScript rules for now and probably include some documentation at the end for linting a mixed workspace. Thanks for the explanation!

**Review by LukasKall — APPROVED** [review]

**Review by mure — APPROVED** [review]

File: typescript.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        /*
+            The following are extension rules that replace core JavaScript rules to support
+            TypeScript.
+        */
+        'brace-style': 'off',
```
[the remark this answers is not in the document]
Comment (rajsite) [reply typescript.js:9]: Going ahead and merging this PR but we can continue this discussion after merge

**Review by rajsite — APPROVED** [review]

File: typescript.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (TrevorKarjanis) [reply typescript.js:9]: It's not difficult to identify the rules as they are documented in a [separate list](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin#extension-rules). When upgrading, changes to these rules can be identified in the changelog under features and breaking changes, [4.2.0 for example](https://github.com/typescript-eslint/typescript-eslint/blob/master/CHANGELOG.md#420-2020-09-21).

[the remark this answers is not in the document]
Comment (rajsite) [reply typescript.js:9]: Could we capture that as a comment in this file? Just from inspection it's not as clear where the list came from

[the remark this answers is not in the document]
Comment (TrevorKarjanis) [reply typescript.js:9]: Yes, we can add it to the configuration file.
